wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,346 @@
+
+    @classmethod
+    def fromhex(cls, hexstr: str) -> Program:
+        return cls.from_bytes(bytes.fromhex(hexstr))
+
+    def __bytes__(self) -> bytes:
+        if self._cached_serialization is None:
+            self._cached_serialization = sexp_to_bytes(self)
+        if not isinstance(self._cached_serialization, bytes):
+            self._cached_serialization = bytes(self._cached_serialization)
+        return self._cached_serialization
+
+    def __int__(self) -> int:
+        v = self.as_int()
+        if v is None:
+            raise ValueError("can't cast pair to int")
+        return v
+
+    def __hash__(self):
+        return id(self)
+
+    @classmethod
+    def int_from_bytes(cls, b: bytes) -> int:
+        return int_from_bytes(b)
+
+    @classmethod
+    def int_to_bytes(cls, i: int) -> bytes:
+        return int_to_bytes(i)
+
+    # high level casting with `.to`
+
+    def __init__(self):
+        self.atom = b""
+        self._pair = None
+        self._unwrapped = self
+        self._unwrapped_pair = None
+        self._cached_serialization = None
+        self._cached_sha256_treehash = None
+
+    @property
+    def pair(self) -> Optional[Tuple["Program", "Program"]]:
+        if self._pair is None and self.atom is None:
+            pair = self._unwrapped_pair
+            self._pair = (self.wrap(pair[0]), self.wrap(pair[1]))
+        return self._pair
+
+    @classmethod
+    def to(cls, v: Any) -> Program:
+        return cls.wrap(to_clvm_object(v, cls.new_atom, cls.new_pair))
+
+    @classmethod
+    def wrap(cls, v: CLVMStorage) -> Program:
+        if isinstance(v, Program):
+            return v
+        o = cls()
+        o.atom = v.atom
+        o._pair = None
+        o._unwrapped = v
+        o._unwrapped_pair = v.pair
+        o._cached_serialization = getattr(v, "_cached_serialization", None)
+        o._cached_sha256_treehash = getattr(v, "_cached_sha256_treehash", None)
+        return o
+
+    # new object creation on the python heap
+
+    @classmethod
+    def new_atom(cls, v: bytes) -> Program:
+        o = cls()
+        o.atom = bytes(v)
+        o._pair = None
+        o._unwrapped_pair = None
+        return o
+
+    @classmethod
+    def new_pair(cls, left: CLVMStorage, right: CLVMStorage) -> Program:
+        o = cls()
+        o.atom = None
+        o._pair = None
+        o._unwrapped_pair = (left, right)
+        return o
+
+    @classmethod
+    def null(cls) -> Program:
+        return NULL_PROGRAM
+
+    # display
+
+    def __str__(self) -> str:
+        s = bytes(self).hex()
+        if len(s) > 76:
+            s = f"{s[:70]}...{s[-6:]}"
+        return s
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({str(self)})"
+
+    def __eq__(self, other) -> bool:
+        try:
+            other_obj = self.to(other)
+        except ValueError:
+            # cast failure
+            return False
+        return self.tree_hash() == other_obj.tree_hash()
```

Comment:
Yes, it would be more efficient to recursively compare leaf nodes, and that code has existed.

However, it's pure python, and tree hashing is generally done in fast native code, so it may end up being faster to do the hash even in the case of it not yet being cached.

I don't expect this to be used very much outside of tests (I believe wallets would typically compare hashes), so porting to native code is probably not warranted.

Reply:
Something we do a lot is `if some_prog == SOME_MOD:` and I would expect we would not have need for either of those hashes so we would end up doing two tree hashes where before we were doing 0.

Reply:
Generally, `SOME_MOD` is going to come from a hex file, which means you'll use `Program.from_bytes`, which by default generates the tree hash for the entire hierarchy (and Coming Soon™ in rust, so it'll be very fast). Same with `some_prog`. Anything that comes from a deserialization rather than building by "hand" (with `Program.to` for example), you'll have to work hard to not pre-generate and cache the tree hashes, which will make these equality comparisons fast.